core/src/consensus/vote_collector.rs

```diff
@@ -69,31 +69,31 @@ impl<M: Message> StepCollector<M> {
     fn insert(&mut self, message: M) -> Option<DoubleVote<M>> {
         // Do nothing when message was seen.
         if self.messages.insert(message.clone()) {
-            if let Some(previous) = self.voted.insert(message.signer_public(), message.clone()) {
+            if let Some(previous) = self.voted.insert(message.signer_index(), message.clone()) {
                 // Bad validator sent a different message.
                 return Some(DoubleVote {
-                    author: public_to_address(&message.signer_public()),
+                    author_index: message.signer_index(),
                     vote_one: previous,
                     vote_two: message,
                 })
             } else {
                 self.block_votes
                     .entry(message.block_hash())
-                    .or_insert_with(HashMap::new)
-                    .insert(message.signature(), message.signer_public());
+                    .or_insert_with(BTreeMap::new)
```

Comment:
You can use [or_default](https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html#method.or_default)